Commit for review

Godzi11a opened the pull request.

VadimGroznov committed:
Commit for review

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+        /// <summary>
+        /// Масштаб изображения - метров в сантиметре экрана
+        /// </summary>
+        public float ScreenScale { get; set; }
```
А в чем разница между **ScreenScale** и, например, **Angle** или **Velocity**?
Почему что-то сделано свойствами, а что-то полями?
По идее, нужно какое-то одноообразие. Ну или поля не должны быть **public**

EALipes replied:
Сделал все свойствами.

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+    {
+        protected Cannon Cannon;
+        protected Projectile Projectile;
+        protected CannonGame Game
```
Как правило, для больших проектов используются соглагения под именованиях. Это нужно, чтобы читать код было проще неподготовленному человеку.
Своим подопечным я разработал такой стандарт:

> Классы всегда именуются с заглавной буквы. Строгих правил на **CamelStyle** или **Name_With_Underscore** нет. Но как по мне, CameStyle преподчтительнее, да и писать меньше
> поля именуются в **_camelStyle**, т.е. имя поля начинается с подчеркивания и строчной буквы (или цифры): `protected Cannon _cannon;`
> свойства именуются с заглавной буквы: `public SomeType CamelStyle { get; set;}` 
> методы именуются со строчной буквы: `void camelStyle()`

Но это дело вкуса и нравов в команде. Если вас в банде два человека - правила можно хоть каждый день менять

EALipes replied:
cameStyle выглядит вот так )
А для свойств и классов используется PascalStyle.
Да, подчеркиваний в C# обычно не используют, но и не запрещают. Видимо, сишникам привычнее.
А вот почему у тебя методы в camelStyle - не понимаю, нигде такого не видел. Только локальные переменные в C# и, видимо, в C.
У меня некоторые переменные могут быть с большой буквы - скорей всего, они превратятся в свойства, и станут публичными, или уже стали.
Насчет постоянства стиля согласен, что полезно.

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+        protected CannonGame Game
+        {
+            get { return (this.Parent as CannonGame); }
+        }
```
Современный C# имеет кое-какой syntax sugar. В частности это свойство лучше записать так:
`protected CannonGame Game => Parent as CannonGame;`
По непроверенным данным, такие конструкции даже оптимизируются компилятором лучше, чем `get { return ...; }` но я не проверял

EALipes replied:
Разумеется, но сначала хотелось бы написать понятно для начинающего.
А потом уже можно освоить и более продвинутые и короткие варианты. Но уже сохранится понимание - что тут делается.

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+        {
+            base.Start();
+
+            Cannon = GetComponent<Cannon>() as Cannon;
```
Я не вполне понял, а почему GetComponent<T> возвращяет не обект типа T ?

EALipes replied:
А вот это клевая идея!

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+            base.Start();
+
+            Cannon = GetComponent<Cannon>() as Cannon;
+            Projectile = GetComponent<Projectile>() as Projectile;
```
По идее пофиг, но в серъезных проектах всегда надо проверять на **null**

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+            // но я еще не придумал куда вставить управление игрой...
+            CheckKeyboardCommands();
+
+            if (Game.Orchestrator.Stopped) return;
```
Не надо экономить строчки. Всегда блок кода размещай на следующе строке после условия. А то в глазах рябить начинает, когда много кода читаешь. Черт разберет, где кончилось условие, а где уже действие

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+
+            projectile.Fallen = false;
+            //projectile
+            Console.WriteLine("Бабах!");
```
Не надо завязываться на **Console**. Для таких вещей лучше завести класс с обощенным набором методов, а их уже менять от реализации к реализации

EALipes replied:
Разумеется. Это артефакты консольного проекта.
Вывод должен будет меняться в зависимости от того, что за приложение.
В консольном - печать в консоль, в графическом - будем рисовать.

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+            //projectile
+            Console.WriteLine("Бабах!");
+
+            projectile.Position.X = Position.Y;
```
По-моему, опечатка, должно ж быть 
```projectile.Position.X = Position.X;```
Кстати, а почему нельзя просто:
`projectile.Position = Position;` ?

EALipes replied:
Исправил.

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+        {
+            //Game = GetParentComponent<Game>() as Game;
+            Game = GetParentComponent<CannonGame>() as CannonGame;
+            if (Game==null) throw new NullReferenceException("Не найден объект игры");
```
Выше уже писал. Разбивай на две строки: 
```
if (условие)
    безжалостноеДействие();
```

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+            time = time + dT;
+
+            // Ускорение свободного падения - 9.81 м/с^2
+            float G = -9.81f; //Направлено вниз, поэтому с минусом
```
Это должна быть константа или переменная объявленная в полях класса, где-то в понятном месте. Типа:
```
public static class Physics
{
  public const float G = -9.81;
}
```
а не переменная, объевленная посреди случайного метода случайного класса

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+
+            // Отрабатываем изменение положения; положение по Y меняется ускоренно.
+            Position.X = Position.X + Velocity.X * dT;
+            Position.Y = Position.Y + (Velocity.Y + dVY/2) * dT ;
```
Вообще странно, что скорость и позиция, это **Vector3**, а не **Vector2**

Godzi11a replied:
Ну и я бы записал так:
```
Position.X += Velocity.X * dT;
Position.Y += (Velocity.Y + dVY/2) * dT ;
```
читать проще

EALipes replied:
Vector3 - более общая штука. А вдруг захотим сделать трехмерку?
Ну, насчет вычислений - да, у тебя красивее. У меня - артефакты придуманной мной ранее структуры, ныне замещенной на Vector3.

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+            // Выводим данные о положении снаряда
+            //Console.WriteLine(DateTime.Now.Minute+":"+DateTime.Now.Second+"."+DateTime.Now.Millisecond);
+            Console.WriteLine("t="+time + "   X="+Position.X.ToString("F2") + "; Y="+Position.Y.ToString("F2") + "  V.Y = "+ Velocity.Y.ToString("F2"));
+            //TODO! Убрать использование Console и сделать обобщенно
```
Правильно писать TODO в комменах так:
`// TODO: что-то, что надо не забыть сделать`
Всякие code-analyser-ы тогда смогут тебе правильно составлять TODO-листы. Иначе они путаются. 
Т.е. пробел перед TODO: и не восклицательный знак, а двоеточие, и потом еще пробел

EALipes replied:
Вроде, так у меня работало. Посмотрим.
P.S. Проверил. Работает
![image](https://github.com/EALipes/MiniUnity/assets/91489965/79fc3d3f-4e41-4073-9df3-34d348288235)

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+            //TODO! Убрать использование Console и сделать обобщенно
+
+            // Если снаряд упал на землю - он останавливается, дальше не летит.
+            if ((Position.Y < 0) & (Velocity.Y<=0))
```
По идее компилятор должен тут варнинг выплюнуть. Ты используешь побитовый AND вместо логического.
Правильно так:
`if (Position.Y < 0 && Velocity.Y <= 0)`

EALipes replied:
Вот тут не понял. Ну, разве что лишнюю проверку тут вызовет. На что тут ругаться?

Godzi11a replied:
Вообще, если ты студенту мастер-класс даёшь, на до такие вещи внимательно проговаривать.
Это довольно большая разница между операторами **"&"** и **"&&"**

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+        // Вероятно, это надо будет просто поставить в конце Scene.Update
+        public void RefreshScreen()
+        {
+            if (OnCallScreenRefresh != null) OnCallScreenRefresh();
```
Сахар: вместо
`if (OnCallScreenRefresh != null) OnCallScreenRefresh();`
можно записать
`OnCallScreenRefresh?.Invoke();`
А в самом новом C# можно будет и вовсе `OnCallScreenRefresh?();` но в Visual Studio это пока не подвезли

EALipes replied:
Пока обходимся без сахара из соображений понятности )

Godzi11a commented on MiniUnity.CannonGame/MiniUnity.CannonGame.cs:
```diff
@@ -1 +1,375 @@
+
+            // Выводим данные о положении снаряда
+            //Console.WriteLine(DateTime.Now.Minute+":"+DateTime.Now.Second+"."+DateTime.Now.Millisecond);
+            Console.WriteLine("t="+time + "   X="+Position.X.ToString("F2") + "; Y="+Position.Y.ToString("F2") + "  V.Y = "+ Velocity.Y.ToString("F2"));
```
Сейчас C# поддерживает F-строки. Можно записать проще:
`Console.WriteLine($"t={time}   X={Position.X:F2}; Y={Position.Y:F2}  V.Y = {Velocity.Y:F2}");`
Ну, конечно, вместо **Console.WriteLine** что-нибудь более обобщенное

Commit:
Переделаны привязки элементов управления в MainForm - сделаны через Binding'и.

Commit:
Фунции поиска объектов в GameObject - теперь возвращают не GameObject, а сразу заказанный тип.

Commit:
Все свойства CannonGame сделаны именно свойствами, а не переменными.

Commit:
Прочие правки

Commit:
Классы вынесены в отдельные файлы

Commit:
Поправлена привязка угла возвышения

Commit:
Немного переструктурированы проекты.
Задание параметров игры всюду сделано через GameParameters.
Пытаемся унифиуировать программы игры на всех платформах.

Commit:
WPF App - Controls in MainWindow

Commit:
GameParameters - поправлено пространство имен
MiniUnity_CannonGame --> MiniUnity.CannonGame

Commit:
Мелкие доработки

Godzi11a commented on MiniUnity.CannonGame/Cannon.cs:
```diff
@@ -0,0 +1,61 @@
+{
+    public class Cannon: GameObject
+    {
+        public Vector3 Position = new Vector3();
```
По идее, это должно быть свойство, а не поле

Godzi11a commented on MiniUnity.CannonGame/Cannon.cs:
```diff
@@ -0,0 +1,61 @@
+            // Отрисовываем снаряд на месте пуска
+            projectile.Update();
+
+            projectile.Fallen = false;
```
Не уверен, что это должно делаться здесь. Как по мне, "упал"/"не упал" это снаряд должен решать, а не пушка

Godzi11a commented on MiniUnity.CannonGame/Cannon.cs:
```diff
@@ -0,0 +1,61 @@
+            var elevationAngleInRadians = elevationAngle * Math.PI / 180;
+
+            projectile.Velocity.X = (float) (velocity * Math.Cos(elevationAngleInRadians));
+            projectile.Velocity.Y = (float) (velocity * Math.Sin(elevationAngleInRadians));
```
Все эти штуки должны выставляться самим снарядом, а то иначе не очень поняяно, зачем вообще было сущности разделать.
У тебя у снаряда есть метод Fall, который его "роняет", мне кажется, должен быть у него и метод Fire(Vector3 positionFrom, float angle), где все эти свойства и будут задаваться

EALipes replied:
Да, сделан такой метод у ядра.

Godzi11a commented on MiniUnity.CannonGame/Projectile.cs:
```diff
@@ -0,0 +1,197 @@
+        public Vector3 Velocity = new Vector3();
+
+        // Отметим момент падения и перестанем сообщать о ранее упавшем снаряде
+        public bool Fallen = false;
```
Все эти поля должны быть readonly, т.к. снаряд сам летит, не надо ему снаружи "помогать"

EALipes replied:
Это использовалось для проверки из сцены - если снаряд уже упал, сцена завершена.
Сделано приватным.
Проверка перенесена в Projectile.

Godzi11a commented on MiniUnity.CannonGame/CannonGame.cs:
```diff
@@ -0,0 +1,37 @@
+        //}
+
+        public Cannon cannon;
+        public Projectile projectile;
```
Это должны быть свойства, а не поля. Ну или поля не должны быть публичными

Godzi11a commented on MiniUnity.CannonGame/CannonScene.cs:
```diff
@@ -0,0 +1,82 @@
+
+            Cannon = GetComponent<Cannon>() as Cannon;
+            if (Cannon==null) throw new NullReferenceException("Не найден объект Cannon");
+            Projectile = GetComponent<Projectile>() as Projectile; 
```
Как я ранее уже писал, GetComponent<T> уже должен возвращать объект типа T, чтобы не делать этих вот кастов всякий раз при вызове

Godzi11a commented on MiniUnity.CannonGame/GameParameters.cs:
```diff
@@ -0,0 +1,144 @@
+            set
+            {
+                _framesPerSec = value;
+                OnPropertyChanged("FramesPerSec");
```
В номом модном C# есть классный оператор nameof, он позволяет такие штуки легче сопровождать:
`OnPropertyChanged(nameof(FramesPerSec));`
Теперь, например, если ты захочешь имя это **FramesPerSec** поменять, то, если ты это сделаешь через Refactor->Rename, оно поменяется и в этом OnPropertyChanged

Godzi11a commented on MiniUnity.CannonGame/GameParameters.cs:
```diff
@@ -0,0 +1,144 @@
+            set
+            {
+                _timeScale = value;
+                OnPropertyChanged("TimeScale");
```
OnPropertyChanged(nameof(TimeScale));

Godzi11a commented on MiniUnity.CannonGame/GameParameters.cs:
```diff
@@ -0,0 +1,144 @@
+            set
+            {
+                _playSound = value;
+                OnPropertyChanged("PlaySound");
```
OnPropertyChanged(nameof(PlaySound));

Godzi11a commented on MiniUnity.CannonGame/GameParameters.cs:
```diff
@@ -0,0 +1,144 @@
+            set
+            {
+                _gameScreenScale = value;
+                OnPropertyChanged("GameScreenScale");
```
OnPropertyChanged(nameof(GameScreenScale));

Godzi11a commented on MiniUnity.CannonGame/GameParameters.cs:
```diff
@@ -0,0 +1,144 @@
+            set
+            {
+                _speed = value;
+                OnPropertyChanged("Speed");
```
OnPropertyChanged(nameof(Speed));

Godzi11a commented on MiniUnity.CannonGame/GameParameters.cs:
```diff
@@ -0,0 +1,144 @@
+
+        //[NotifyPropertyChangedInvocator]
+        //protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null)
+        protected virtual void OnPropertyChanged(string propertyName = null)
```
Я очень плохо себе представляю, зачем может понадобиться PropertyChanged для **null**

EALipes replied:
Это Решарпер сделал метод. 
См. закомментированную строчку выше. Там используется атрибут [CallerMemberName],
я его пока убрал.
Но по идее - тут  мы могли бы не задавать имена свойств, т.е. в сеттерах вызывать просто PropertyChanged(),
а имя свойства подставлялось бы само.
Я, правда, не очень понимаю, это компилятор его тут вычислит и подставит, или кто?

А вообще, 
"Событие PropertyChanged может указать, что все свойства объекта изменились, используя null или String.Empty в качестве имени свойства" - так говорил ~~Заратустра~~ Microsoft.
Еще я, помнится, использовал это, когда в контроле выводил объект через его собственный ToString().

Godzi11a commented on MiniUnity.CannonGame/Projectile.cs:
```diff
@@ -0,0 +1,197 @@
+            //TODO! Убрать использование Console и сделать обобщенно
+            Console.WriteLine("Шлёп!");
+            Fallen = true;
+            Scene.IsOver = true;
```
Не надо так делать. А вдруг ты захочешь, чтобы снарядов у тебя было больше одного? 
Сцена сама должна решать, когда она **IsOver**. Мало того, она у тебя это и так делает в своём **Update**

Godzi11a commented on MiniUnity.CannonGame/Projectile.cs:
```diff
@@ -0,0 +1,197 @@
+        {
+            // Похоже, вообще не требуется.
+            // Вместо этого форма установит в OnCallScreenRefresh свой this.Refresh()
+        }
```
Не самое лучшее, по-моему, решение - иметь все возможные реализации Redraw в одном классе. Это всё гораздо лучше реализуется через наследование и переопределение метода

EALipes replied:
Ты, безусловно, прав.
Я тут показываю как можно, потом будем делать как нужно.

Godzi11a commented on MiniUnity/Game.cs:
```diff
@@ -0,0 +1,121 @@
+            if (Scene == null) throw new NullReferenceException("Не определено отыгрываемых сцен");
+
+            game.IsOver = false;
+            scene.IsOver = false;
```
Эти поля должны сбрасываться самими объктами. Например, при вызове Start у них.
Не надо снаружи менять свойства объекта, потом задолбаешься искать, кто тебе в неверный стейт какое-то свойство перевёл

EALipes replied:
```
        public void Play(Game game, Scene scene)
        {
            ...
            ....  
            Start();
            //// Вместо всего этого:
            //game.IsOver = false;
            //scene.IsOver = false;
            //Orchestrator.Start();
            //Scene.Start();
```

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+        public void RemoveComponent(GameObject gameObject)
+        {
+            Children.Remove(gameObject);
+            gameObject.Parent = null;
```
Опасно. Надо убедиться сначала, что этот объект именно наш детёныш:
```
if (Children.Remove(gameObject))
    gameObject.Parent = null;
```

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+        public void AddComponent(GameObject gameObject)
+        {
+            Children.Add(gameObject);
+            gameObject.Parent = this;
```
Опасно. Может возникнуть ситуация, что объкт будет в списках Children у разных Parent-ов.
Надо сначала выкинуть его из списка старого родителя:
```
gameObject.Parent?.RemoveComponent(gameObject);
Children.Add(gameObject);
gameObject.Parent = this;

```

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+            var result = Children?
+                .Where(b => (b.GetType() == typeof(T)) || (b.GetType().IsSubclassOf(typeof(T))))
+                .Select(b=> b as T)
+                .ToList();
```
Проверку в Where можно упростить:
`.Where(b => b is T)`

Godzi11a replied:
Ну и чтобы два раза таблицу RTTI не теребить, я бы сделал так:
`var result = Children.Select(b=> b as T).Where(x => x != null).ToList();`
хотя это может уже и вкусовщина...

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+            where T:GameObject
+        {
+            // Тоже LINQ
+            var result = Children?
```
Children не может быть **null**, он у тебя инициализируется пустым листом при создании экземпляра класса

EALipes replied:
Ну, перестраховался. От этого будет хуже?

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+        {
+            // Мне лень честно писать поиск первого подходящего элемента такого типа в списке,
+            // поэтому я использовал LINQ, хотя это заклинание проходят только на старших курсах Хогвартса )))
+            var result = Children?.FirstOrDefault(b => (b.GetType() == typeof(T)) || (b.GetType().IsSubclassOf(typeof(T))) );
```
Можно упростить:
`return Children.FirstOrDefault(b => b is T) as T;`

EALipes replied:
О, оно тут есть и работает?
А я чего-то как-то его искал и не нашел.

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+                .Where(b => (b.GetType() == typeof(T)) || (b.GetType().IsSubclassOf(typeof(T))))
+                .Select(b => b as T)
+                .ToList();
+            return result;
```
Дублирование кода. Ты то же самое делаешь в **GetComponents<T>**
Я бы сделал так:
```
public ICollection<T> GetChildrenOfType<T>() where T : class
{
	return Children.Where(x => x is T).Select(x => x as T).ToList();
}

public ICollection<T> GetComponents<T>() where T : GameObject => GetChildrenOfType<T>();
public ICollection<T> GetBehaviors<T>() where T : BehaviorComponent => GetChildrenOfType<T>();
```

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+            foreach (GameObject c in Children)
+            {
+                c.Start();
+                if (OnStart != null) OnStart(this);
```
Похоже на ошибку. Ты так вызовешь OnStart(this) столько раз, сколько элементов в Children.
По идее, надо так:
```
foreach (GameObject c in Children)
    c.Start();

OnStart?.Invoke(this);
```

Godzi11a commented on MiniUnity/GameObject.cs:
```diff
@@ -0,0 +1,160 @@
+            foreach (GameObject c in Children)
+            {
+                c.Update();
+                if (OnUpdate != null) OnUpdate(this);
```
То же самое, ты так OnUpdate(this); вызовешь несколько раз (сколько детей, столько и раз)

Godzi11a commented on MiniUnity/UpdateOrchestrator.cs:
```diff
@@ -0,0 +1,194 @@
+        public float GameTimeScale
+        {
+            get { return Clock.GameTimeScale;}
+            set { Clock.GameTimeScale = value; }
```
А если часы не созданы?

EALipes replied:
По идее, часы должны быть.
Если их нет - будет исключение NullReference... или какеготам...
Что предлагаешь сделать?

Godzi11a commented on MiniUnity/UpdateOrchestrator.cs:
```diff
@@ -0,0 +1,194 @@
+
+        public void Start()
+        {
+            CurrentUpdateTime = Now();
```
А если часы не созданы?

По идее, часы как раз и надо создавать в самом начале этого метода Start

EALipes replied:
Так они у меня в конструкторе создаются.

Godzi11a commented on MiniUnity_Cannon_ConsoleApp/Program.cs:
```diff
@@ -0,0 +1,121 @@
+            var repeat = (s.KeyChar == 'y') | (s.KeyChar == 'Y') | (s.KeyChar == 'Д') | (s.KeyChar == 'д') |
+                         (s.Key == ConsoleKey.Enter);
+            if (repeat) return true;
+            return false;
```
Длинные if-ы довольно сложно саппортить. Я бы сделал через switch-case:
```
switch (s.KeyChar)
{
    case 'y':
    case 'Y':
    case 'д':
    case 'Д':
        return true;
}
return s.Key == ConsoleKey.Enter;
```

Godzi11a commented on MiniUnity_Cannon_ConsoleApp/Program.cs:
```diff
@@ -0,0 +1,121 @@
+                var s = Console.ReadLine();
+                if (String.IsNullOrWhiteSpace(s) && defaultValue.HasValue)
+                    return defaultValue.Value;
+                if (float.TryParse(s, out result)) return result;
```
Современный C# позволяет объявлять **out** переменные **in-place**:
```
if (float.TryParse(s, out float result))
    return result;
```

EALipes replied:
Удобно. В следующий раз так и сделаю, если не забуду ) Спасибо.

Godzi11a commented on MiniUnity_Cannon_DesktopApp/MainForm.cs:
```diff
@@ -0,0 +1,283 @@
+                // Звук выстрела (действие начинается после его окончания)
+                if (game.PlaySound)
+                    //SoundPlayerGunFired.Play();
+                    new SoundPlayer( Resources.CannonFiredAndProjectileFlies).Play();
```
Вообще, мне кажется, что проигрыванием звука должен заниматься объект **game**

EALipes replied:
Может быть. Пока не придумал, как это сделать лучше.
На настоящий момент я как раз убрал издавание звуков из игры, передал это самим объектам - пушке и ядру.

Godzi11a commented on MiniUnity_Cannon_DesktopApp/MainForm.cs:
```diff
@@ -0,0 +1,283 @@
+
+        private void VelocityChanged_to_VelocityEdit(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "Velocity")
```
У тебя вроде нет такого свойства, у тебя есть **Speed**

Godzi11a replied:
Поэтому как раз, лучше и использовать **nameof**:
```
if (sender is GameParameters prms && e.PropertyName == nameof(GameParameters.Speed))
    VelocityEdit.Value = (decimal)prms.Speed;
```

Godzi11a commented on MiniUnity_Cannon_DesktopApp/MainForm.cs:
```diff
@@ -0,0 +1,283 @@
+        private void AngleChanged_to_AngleEdit(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "Angle")
+                AngleEdit.Value = (decimal)(sender as GameParameters)?.Angle;
```
См. комментарий к VelocityChanged_to_VelocityEdit

Godzi11a commented on MiniUnity_Cannon_WpfApp/MainWindow.xaml.cs:
```diff
@@ -0,0 +1,77 @@
+            var b2 = new Binding("Speed");
+            b2.Source = gameParams;
+            VelocitySlider.Minimum = 0;
+            VelocitySlider.Maximum = 2000;
```
Это ж можно задать в XAML

EALipes replied:
Тут я хотел показать, что привязки работают так же, как и в Winforms.
Ну, честно говоря, тут весь модуль коряво написан, и я до сих пор не могу заставить тут ядро нормально отображаться. Оказывается, WPF я основательно позабыл, а с отображением графики в нем вообще не работал - я ж в нем цифры и ссылки редактировал, офисные приложения.
Сейчас ищу, кто поможет с летящим ядром )

Godzi11a commented on MiniUnity_Cannon_WpfApp/MainWindow.xaml.cs:
```diff
@@ -0,0 +1,77 @@
+            var bA2 = new Binding("Angle");
+            bA2.Source = gameParams;
+            AngleSlider.Minimum = 0;
+            AngleSlider.Maximum = 90;
```
Это тоже можно задать в XAML, для общности

Godzi11a commented on MiniUnity_Cannon_WpfApp/MainWindow.xaml.cs:
```diff
@@ -0,0 +1,77 @@
+            bA2.Source = gameParams;
+            AngleSlider.Minimum = 0;
+            AngleSlider.Maximum = 90;
+            AngleSlider.SetBinding(Slider.ValueProperty, bA2);
```
Бинды, кстати, тоже можно задать в XAML, там оно проще и читать легче

Commit:
Правка модели ядра: Cannon и Projectile должны издавать звук сами, этим не должно заниматься приложение!

Commit:
Попытка отрисовки летящего ядра в WPF App. Пока не получается.

Commit:
Мелкие правки по ресурсам.

Commit:
Мелкие правки

Commit:
Переработка отрисовки в Console, WinForms, WPF - начато, недокончено.
Методы для GameObject, Game.

Commit:
Переработка отрисовки в Console, WinForms -  заработало.
Обе отрисовки реализованы методами GameObject.
Т.е. он изменяется каждый раз, когда потребуется новая отрисовка.
Плюс к его сборке добавляются ссылки на новые сборки.

Commit:
Методы отрисовки Draw() и Draw_OnWinFormsPaintEvent() сделаны protected, чтоб их не вызывали напрямую, а только перекрывали.
Для прямого вызова следует использовать только RefreshDraw().

Commit:
Мелкие правки по замечаниям.

Commit:
Причёсана функция обновления при полете снаряда.
